### ISSUE TITLE
a11y: announce changes in the avatar group

### ISSUE THREAD

**Issue by web-padawan** [vaadin/vaadin-avatar#44]
From https://github.com/vaadin/vaadin-avatar/issues/10#issuecomment-657055930

> I was thinking about adding aria-live to <vaadin-avatar-group> as well.

Consider using `iron-a11y-announcer` to announce changes in the group.  It's a bit unclear how this should be done.
Most likely, it would make sense to announce when a user is added or removed to the group.

@knoobie would you like to clarify your expectations here?

**Comment by knoobie** [comment]
I have two opinions:

- if `aria-live`works, we could rely on `aria-label` on `vaadin-avatar-group` to update the amount of users online (+1 / -1) and the `vaadin-avatar` with the correct title attribute gets automatically announced if added or removed.
- if `aria-live` doesn't work, it would be good to throw events like `VaadinAvatarAdded` and `VaadinAvatarRemoved` which are announced with `iron-a11y-announcer` like "User {{title}} is online now." / "User {{title}} is offline now." (i18n of course)


Update: More informations, if `aria-live` can be used:

![image](https://user-images.githubusercontent.com/3968629/87394830-9b619100-c5b0-11ea-8af0-14316e81b5a3.png)
Source: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions

**Comment by knoobie** [comment]
I read a little bit about `aria-live`- I would propose to try this road as first step! 

You can find more informations here: https://developers.google.com/web/fundamentals/accessibility/semantics-aria/hiding-and-updating-content#aria-live 

The first issue, we have to test is, that is won't happen: 

> First, your aria-live region should probably be set in the initial page load. This is not a hard-and-fast rule, but if you're having difficulty with an aria-live region, this might be the issue.

A really common problem with aria-live.. Would be cool to have a demo application that "enables" and "disables"  the `vaadin-avatar-group`. (adding / removing from the dom) 

Additionally, I'm not 100% sure (has to be tested), if we can place the `aria-label` onto `<vaadin-avatar-group>` and the changes are announced as well, or if we have to create a screenreader-only div/span inside the `<vaadin-avatar-group>` (A  `<vaadin-sr-only>` sounds like an interesting idea for me?) to announce the changes of +1/-1 user. Reading the WAI ARIA spec, I think we have to go for a screenreader-only text INSIDE of `<vaadin-avatar-group>`

About the attributes, we would need on `vaadin-avatar-group`:
- aria-live="polite"
- role="marquee" (https://www.digitala11y.com/marquee-role/)
- aria-atomic="false" (If aria-atomic is explicitly set to false, assistive technologies will stop searching up the ancestor chain and present only the changed node to the user.)
- aria-relevant="additions text removals" (additions and removals are obvious; text is "Text content or a text alternative is added to any descendant in the accessibility tree of the live region.")

https://www.w3.org/TR/2017/REC-wai-aria-1.1-20171214/#aria-relevant

**Comment by web-padawan** [comment]
Thanks for the detailed research. 

The AvatarGroup is supposed to be mainly used with Flow, and more specifically [Collaboration Engine](https://vaadin.com/collaboration) so in most of cases the avatar groups won't be there on initial page load. Which means we probably need a different approach.

Would `iron-a11y-announcer` be enough? This is the way how the user starting to edit the field is announced:

```js
const msg = `${user.name} started editing`;
this.dispatchEvent(new CustomEvent('iron-announce', {
  bubbles: true,
  composed: true,
  detail: {
    text: label ? `${msg} ${label}` : msg
  }
}));
```

The typical scenario as I see it looks like this:

1. User X joined (announcement from `vaadin-avatar-group`)
2. User X started editing field (announcement from the `vaadin-field-highlighter` component)

The same, but in the reverse order, should probably happen when user stops editing and leaves.

**Comment by knoobie** [comment]
If you already have good experience with `iron-a11y-announcer` or use it somewhere else it should work as well! It sounds good to use one approach in all components, instead of creating a second way of doing things. 

Tomorrow, I'm gonna ask our screenreader user if he can work with it, using the demo provided by polymer here:  https://www.webcomponents.org/element/@polymer/iron-a11y-announcer

**Comment by knoobie** [comment]
I've got hold on our screenreader user - he can hear the announcements of iron-a11y-announcer thrown by the polymer demo. Using JAWS (2018, old I know) and NVDA. 

He got some problems clicking the other announce "buttons" with his braille keyboard. Only the first one could be interacted with. I helped him with visually clicking the others - at least all announcend it.

I'm gonna update this post once I know what they did with them (typing on mobile) 

Edit: Well.. using `aria-hidden` is totally fucked up.. That explains it. 
![image](https://user-images.githubusercontent.com/3968629/87659256-97bf3d00-c75d-11ea-92f3-3d7cb43fee32.png)

**Comment by web-padawan** [comment]
Thank you for testing. I'm going to make a prototype using `iron-a11y-announcer` tomorrow and will do it in #49

**Comment by web-padawan** [comment]
Sorry about the delay. I was on vacation and then we had other blockers. Will get back to this today.

**Comment by web-padawan** [comment]
@knoobie I have submitted a PR with a demo that can be used to test dynamic changes in avatar group: https://github.com/vaadin/vaadin-avatar/pull/63

Would you like to test this approach when you have time? I only tried it in VoiceOver and there it seems to work.

**Comment by knoobie** [comment]
Sadly I have no way to share the current snapshot with my colleague (He is in a secured environment). Once you release the next public available demo, I can talk with him!

**Comment by web-padawan** [comment]
Thank you for letting me know. I will add finish the PR and then will let you know when public demo is available.

**Comment by web-padawan** [comment]
@knoobie the change that I linked above is now merged and the demo is available here:
https://cdn.vaadin.com/vaadin-avatar/1.0.0-alpha8/demo/#avatar-advanced-demos

I would really appreciate your feedback. So far I have only tested it using VoiceOver on Safari and Chrome.

**Comment by knoobie** [comment]
@web-padawan I tested it with our colleague. It worked for him in Firefox with NVDA as well! :+1: (his braille keyboard was notified and the sound output worked)

**Comment by web-padawan** [comment]
Thank you, that's nice to hear. So I'm closing this issue as fixed.
